frontend/src/metabase/visualizations/lib/settings/graph.js

```diff
@@ -151,35 +151,56 @@ export const GRAPH_DATA_SETTINGS = {
     dashboard: false,
     useRawSeries: true,
   },
+  "graph.series_order_dimension": {
+    getValue: (_series, settings) => settings["graph.dimensions"][1],
+    readDependencies: ["graph.series_order"],
+  },
   "graph.series_order": {
     section: t`Data`,
     widget: ChartSettingOrderedSimple,
     marginBottom: "1rem",
-    isValid: (series, settings) => {
+
+    getValue: (series, settings) => {
+      const seriesKeys = series.map(s => keyForSingleSeries(s));
       const seriesOrder = settings["graph.series_order"];
+      const seriesOrderDimension = settings["graph.series_order_dimension"];
+      const currentDimension = settings["graph.dimensions"][1];
+
+      const generateDefault = keys => {
+        return keys.map(key => ({
+          key,
+          name: key,
+          enabled: true,
+        }));
+      };
+
+      const removeMissingOrder = (keys, order) =>
+        order.filter(o => keys.includes(o.key));
+      const newKeys = (keys, order) =>
+        keys.filter(key => !order.find(o => o.key === key));
 
-      if (!seriesOrder || !_.isArray(seriesOrder)) {
-        return false;
+      if (
+        !seriesOrder ||
+        !_.isArray(seriesOrder) ||
+        seriesOrderDimension !== currentDimension
+      ) {
+        return generateDefault(seriesKeys);
       }
 
-      return seriesOrder.length === series.length;
-    },
-    getDefault: series => {
-      const keys = series.map(s => keyForSingleSeries(s));
-      return keys.map((key, index) => ({
-        name: key,
-        originalIndex: index,
-        enabled: true,
-      }));
+      return [
+        ...removeMissingOrder(seriesKeys, seriesOrder),
+        ...generateDefault(newKeys(seriesKeys, seriesOrder)),
```

Comment:
I think it's reasonable that we append new keys that aren't in the `graph.series_order` last. But it's worth discussing if this is what we all agree with.
#### Initial question has only 3 columns due to the filter
![Screenshot 2022-11-22 at 2 59 02 PM](https://user-images.githubusercontent.com/1937582/203258023-6b11cb27-55f8-4317-952a-bda40a27fa9b.png)

#### After removing the filter, it will show missing columns last
![Screenshot 2022-11-22 at 2 59 10 PM](https://user-images.githubusercontent.com/1937582/203258121-6f61e3d9-7332-48d1-b12a-5398e5aa8a42.png)

Reply:
I chatted with @cdeweyx about this yesterday, and this is the behavior we landed on for the time being. Certainly room for other UX if we find it's necessary though!